code/modules/mob/living/carbon/xenomorph/castes/gorger/abilities_gorger.dm

```diff
@@ -59,78 +168,69 @@
 			if(!silent)
-				to_chat(X, span_notice("It is beyond our reach, we must be close and our way must be clear."))
+				to_chat(owner_xeno, span_notice("It is beyond our reach, we must be close and our way must be clear."))
 			return FALSE
-		if(X.plasma_stored < GORGER_REJUVENATE_ALLY_COST)
+		if(owner_xeno.plasma_stored < GORGER_REJUVENATE_ALLY_COST)
 			if(!silent)
-				to_chat(X, span_notice("We need [GORGER_REJUVENATE_ALLY_COST - X.plasma_stored]u more blood to restore a sister."))
+				to_chat(owner_xeno, span_notice("We need [GORGER_REJUVENATE_ALLY_COST - owner_xeno.plasma_stored]u more blood to restore a sister."))
 			return FALSE
 		if(isdead(target))
 			if(!silent)
-				to_chat(X, span_notice("We can only help living sisters."))
+				to_chat(owner_xeno, span_notice("We can only help living sisters."))
 			return FALSE
-		if(!do_mob(X, target, 1 SECONDS, BUSY_ICON_FRIENDLY, BUSY_ICON_MEDICAL))
+		if(!do_mob(owner_xeno, target, 1 SECONDS, BUSY_ICON_FRIENDLY, BUSY_ICON_MEDICAL))
 			return FALSE
 		return TRUE
 
-/datum/action/xeno_action/activable/rejuvenate/use_ability(atom/A)
-	var/mob/living/carbon/xenomorph/X = owner
-	if(ishuman(A))
-		var/mob/living/carbon/human/target = A
-		while(target.blood_volume > GORGER_REJUVENATE_BLOOD_DRAIN && X.health < X.maxHealth && do_after(X, 2 SECONDS, TRUE, A, BUSY_ICON_HOSTILE))
-			X.heal_wounds(2.2, TRUE)
-			X.adjust_sunder(-1.5)
-			target.blood_volume -= GORGER_REJUVENATE_BLOOD_DRAIN
-		to_chat(X, span_notice("We feel fully restored."))
-	else if(A == X)
-		X.use_plasma(GORGER_REJUVENATE_SELF_COST)
-		X.apply_status_effect(/datum/status_effect/xeno_rejuvenate, GORGER_REJUVENATE_SELF_DURATION)
-		to_chat(X, span_notice("We tap into our reserves for nourishment."))
+/datum/action/xeno_action/activable/rejuvenate/use_ability(atom/target)
+	var/mob/living/carbon/xenomorph/owner_xeno = owner
+	if(target == owner_xeno)
+		owner_xeno.use_plasma(GORGER_REJUVENATE_SELF_COST)
+		owner_xeno.apply_status_effect(/datum/status_effect/xeno_rejuvenate, GORGER_REJUVENATE_SELF_DURATION)
+		to_chat(owner_xeno, span_notice("We tap into our reserves for nourishment."))
 		add_cooldown()
 	else
-		X.use_plasma(GORGER_REJUVENATE_ALLY_COST)
-		TIMER_COOLDOWN_START(X, REJUVENATE_ALLY, GORGER_REJUVENATE_ALLY_COOLDOWN)
-		var/mob/living/carbon/xenomorph/target_xeno = A
-		target_xeno.adjustBruteLoss(-target_xeno.maxHealth*0.3)
-		target_xeno.adjustFireLoss(-target_xeno.maxHealth*0.3)
+		owner_xeno.use_plasma(GORGER_REJUVENATE_ALLY_COST)
+		TIMER_COOLDOWN_START(owner_xeno, REJUVENATE_ALLY, GORGER_REJUVENATE_ALLY_COOLDOWN)
+		var/mob/living/carbon/xenomorph/target_xeno = target
+		target_xeno.adjustBruteLoss(-target_xeno.maxHealth * GORGER_REJUVENATE_ALLY_PERCENTAGE)
+		target_xeno.adjustFireLoss(-target_xeno.maxHealth * GORGER_REJUVENATE_ALLY_PERCENTAGE)
 	succeed_activate()
 
+/datum/action/xeno_action/activable/rejuvenate/ai_should_use(atom/target)
+	// no healing non-xeno
+	if(!isxeno(target))
```

Comment:
This should be a self heal as well ? Since it's never targeting itself

Reply:
What do I need to change for it to also heal itself?

Reply:
if(owner_health && can_use_ability)
    use_ability(owner)

something like that

Reply:
added